### PR TITLE
refacto previsualization

### PR DESCRIPTION
**Pull request by mathieudpnt**
I fixed the adjust_spectro located in src/utils_datarmor, it should run smoothly provided the jupyter hub session runs with enough ressources
> By default, we recommend to use Datarmor - Jupyterlab - 8 cores, 32 GB RAM, 2 hours

### CODE REVIEW

File: src/utils_datarmor.py [hunk 1]
```diff
@@ -105,59 +53,79 @@ def adjust_spectro(
...
+    for _ in range(number_adjustment_spectrogram):
+        random_idx = randint(0, len(origin_files) - 1)
 
-        if os.path.exists(temp_adjustment_output_dir):
-            shutil.rmtree(temp_adjustment_output_dir)
+        selected_ts_beg = pd.Timestamp(
+            file_metadata["timestamp"].iloc[random_idx]
+        ).timestamp()
+        selected_ts_end = selected_ts_beg + file_metadata["duration"].iloc[random_idx]
+        tz_data = pd.Timestamp(file_metadata["timestamp"].iloc[random_idx]).tz
```
Comment (Gautzilla) [inline src/utils_datarmor.py:68]: Not a big deal, but pandas' `sample` might help you here:
```py
random_file = file_metadata.sample() # Gets a random row in file_metadata
```

Plus, you could work with Timestamps and Timedeltas to get rid of the timezone dependency and of the `Timestamp(...)`/`.timestamp()` conversions! 

```py
file_start = Timestamp(random_file["timestamp"].values[0])
file_stop = file_start + Timedelta(seconds = random_file["duration"].values[0])
spectro_start = file_start + Timedelta(seconds = randint(0, int((file_end-file_start).seconds - spectro_duration)))
```

(I haven't tested it in situ tho 🥸)

**Review by Gautzilla — COMMENTED** [review]
Just a small remark on how you could keep the computing of the reshape start/stop simpler!

**Comment by Gautzilla** [comment]
When trying to call `generate_spectro()`, I have the following error:
```py
AssertionError: 'datetime_begin' must be either 'None' or a datetime, None not a valid value
```

I didn't dig into what you wrote, but I guess it comes from the removing of the `validate_datetime` function?

**Comment by Gautzilla** [comment]
Shouldn't the temporary folder that holdsthe adjustment spectrograms be deleted after the display rather than before displaying new spectrograms?

After plotting some adjustment spectros, I'm left with this structure:

image

File: src/utils_datarmor.py [hunk 2]
```diff
@@ -105,59 +53,88 @@ def adjust_spectro(
...
-        files_adjust = [temp_adjustment_output_dir + "/" + ff for ff in file_list]
+        if all([f in file_metadata["filename"].values for f in file_list]):
+            file_metadata = file_metadata[file_metadata["filename"].isin(file_list)]
+        else:
+            files_not_in_file_metadata = [
+                f for f in file_list if f not in file_metadata["filename"].values
+            ]
+            raise ValueError(
+                f"{files_not_in_file_metadata} not found in file_metadata file."
+            )
```
Comment (Gautzilla) [inline src/utils_datarmor.py:67]: We might clean that out later, but that's no big deal! 🥸

File: src/utils_datarmor.py [hunk 3]
```diff
@@ -170,14 +147,30 @@ def generate_spectro(
...
+
+    if not datetime_end:
+        datetime_end = file_metadata["timestamp"].iloc[-1] + pd.Timedelta(
+            file_metadata["duration"].iloc[-1], unit="s"
+        )
+    else:
+        try:
+            datetime_end = pd.Timestamp(datetime_end)
+        except Exception as e:
+            raise ValueError(f"'datetime_end' not a valid datetime: {e}")
```
Comment (Gautzilla) [inline src/utils_datarmor.py:173]: Lots of lines just to clamp the timestamps, I'll clean that with OSEkit utils some day

File: src/utils_datarmor.py [hunk 4]
```diff
@@ -233,7 +226,7 @@ def generate_spectro(
             logdir=log_dir,
         )
 
-        jobfiles.append(jobfile)
+        job_files.append(job_file)
```
Comment (Gautzilla) [inline src/utils_datarmor.py:229]: ‼️‼️‼️
Have you tested that? I don't see `job_files` being declared anywhere, doesn't it lead to a bug ??
‼️‼️‼️

Comment (Gautzilla) [reply src/utils_datarmor.py:229]: I just tested it, it raises an error:

```py
generate_spectro(
    dataset=dataset,
    path_osmose_dataset=path_osmose_dataset,
    write_datasets_csv_for_aplose = True,
    overwrite=True,
    save_matrix=save_matrix,
    save_welch=save_welch,
    datetime_begin=datetime_begin,
    datetime_end=datetime_end,
)
```

```py
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[4], line 9
      5 datetime_end = None
      7 dataset.batch_number = 1
----> 9 generate_spectro(
     10     dataset=dataset,
     11     path_osmose_dataset=path_osmose_dataset,
     12     write_datasets_csv_for_aplose = True,
     13     overwrite=True,
     14     save_matrix=save_matrix,
     15     save_welch=save_welch,
     16     datetime_begin=datetime_begin,
     17     datetime_end=datetime_end,
     18 )

File ~/datarmor-toolkit/notebooks/../src/utils_datarmor.py:229, in generate_spectro(dataset, path_osmose_dataset, write_datasets_csv_for_aplose, overwrite, save_matrix, save_welch, datetime_begin, datetime_end)
    207     last_file_index = first_file_index + batch_sizes[batch]
    209     job_file = dataset.jb.build_job_file(
    210         script_path=Path(os.path.abspath("../src")).joinpath(
    211             "qsub_spectrogram_generator_pkg.py"
   (...)
    226         logdir=log_dir,
    227     )
--> 229     job_files.append(job_file)
    231 if hasattr(dataset, "pending_jobs"):
    232     pending_jobs = [
    233         jobid
    234         for jobid in dataset.pending_jobs
    235         if b"finished"
    236         not in subprocess.run(["qstat", jobid], capture_output=True).stderr
    237     ]

NameError: name 'job_files' is not defined
```

Comment (Gautzilla) [reply src/utils_datarmor.py:229]: It seems that you just omitted to replace `jobfiles` with `job_files` in its declaration line `195`

**Comment by mathieudpnt** [comment]
> Shouldn't the temporary folder that holdsthe adjustment spectrograms be deleted after the display rather than before displaying new spectrograms?
> 
> After plotting some adjustment spectros, I'm left with this structure:
> 
> image

@Gautzilla The temporary folder will be deleted when you compute the actual spectrograms of the dataset so this is normal behavior to me 🥸

File: src/utils_datarmor.py [hunk 5]
```diff
@@ -1,95 +1,43 @@
...
-        dataset._get_original_after_build().joinpath("metadata.csv"), header=0
-    )
-    orig_dura = orig_metadata["audio_file_origin_duration"][0]
-    orig_sr = orig_metadata["origin_sr"][0]
-
-    origin_files = []
-    for ext in SUPPORTED_AUDIO_FORMAT:
-        origin_files_ext = list(dataset.audio_path.glob(f"*{ext}"))
-        [origin_files.append(f) for f in origin_files_ext]
+    origin_files = get_all_audio_files(dataset.audio_path)
```
[the remark this answers is not in the document]
Comment (mathieudpnt) [reply src/utils_datarmor.py:40]: will be deleted

File: src/utils_datarmor.py [hunk 6]
```diff
@@ -105,59 +53,88 @@ def adjust_spectro(
...
+        random_file = file_metadata.sample()  # Gets a random row in file_metadata
 
-        if os.path.exists(temp_adjustment_output_dir):
-            shutil.rmtree(temp_adjustment_output_dir)
+        file_start = pd.Timestamp(random_file["timestamp"].values[0])
+        file_end = file_start + pd.Timedelta(seconds=random_file["duration"].values[0])
+
+        spectro_start = file_start + pd.Timedelta(
+            seconds=randint(
+                0, int((file_end - file_start).seconds - dataset.spectro_duration)
```
[the remark this answers is not in the document]
Comment (mathieudpnt) [reply src/utils_datarmor.py:80]: defo a better implementation

File: src/utils_datarmor.py (hunk 3 above)
Comment (mathieudpnt) [reply src/utils_datarmor.py:173]: yes i did a dirty snippet as I new it would be dealt with in OSEkit later

**Review by Gautzilla — APPROVED** [review]

**Pull request merged**